### PR TITLE
Greensteel eco sync

### PR DESCRIPTION
**Pull request by jaredthomas68**
# Update greenheart to more closely match the original framework from the GreenSteel project.

This PR brings in many changes to align the greenheart framework with the framework from the original analysis in the GreenSteel project. It allows users to make adjustments so that most components can be placed either onshore or offshore. A rudimentary hydrogen dispatch estimation has been included, but is only calculated if hydrogen storage is being sized based on the desired load or average hydrogen production.

## Related issue

#281 
#282 
#283 

## Impacted areas of the software

Primarily `greenheart` but a few changes in HOPP and package files as well.

### CODE REVIEW

File: greenheart/simulation/technologies/hydrogen/electrolysis/PEM_H2_LT_electrolyzer_Clusters.py [hunk 1]
```diff
@@ -63,31 +63,21 @@ class PEM_H2_Clusters:
     """
     #Remove: estimate_lifetime_capacity_factor
     #Remove: make_lifetime_performance_df_all_opt [x]
-    def __init__(self, cluster_size_mw, plant_life, user_defined_EOL_percent_eff_loss, eol_eff_percent_loss=[],user_defined_eff = False,rated_eff_kWh_pr_kg=[],include_degradation_penalty=True,turndown_ratio=0.1,dt=3600):
+    def __init__(self, cluster_size_mw, plant_life, eol_eff_percent_loss=10,uptime_hours_until_eol = 77600, include_degradation_penalty=True,turndown_ratio=0.1,dt=3600):
```
Comment (bayc) [inline greenheart/simulation/technologies/hydrogen/electrolysis/PEM_H2_LT_electrolyzer_Clusters.py:66]: Why the changes to the default values here?

Comment (bayc) [reply greenheart/simulation/technologies/hydrogen/electrolysis/PEM_H2_LT_electrolyzer_Clusters.py:66]: Nvm, I see the removal of the else code below that had these defaults in it.

File: greenheart/simulation/technologies/hydrogen/electrolysis/run_h2_PEM.py [hunk 2]
```diff
@@ -99,56 +104,41 @@ def run_h2_PEM(electrical_generation_timeseries, electrolyzer_size,
...
-                  'avg_stack_life_hrs':
-                  average_stack_life_hrs,
-                  # h2_tot.loc['Avg [hrs] until Replacement Per Stack'].mean(),
-                  'Rated kWh/kg-H2':rated_kWh_pr_kg,
-                  'average_operational_time [hrs]':
-                  average_uptime_hr,
-                  'new_H2_Results':new_H2_Results,
-                  'Performance Schedules':pd.DataFrame(annual_avg_performance),
-                  }
+   # H2_Results = {
```
Comment (bayc) [inline greenheart/simulation/technologies/hydrogen/electrolysis/run_h2_PEM.py:123]: Can this be removed now?

File: greenheart/simulation/technologies/steel/steel.py [hunk 3]
```diff
@@ -759,8 +849,108 @@ def run_steel_finance_model(
     summary = pf.get_summary_vals()
     price_breakdown = pf.get_cost_breakdown()
 
+    if config.save_plots or config.show_plots:
```
Comment (bayc) [inline greenheart/simulation/technologies/steel/steel.py:852]: The output processing/manipulation/plotting should be moved elsewhere, in my opinion. That falls more into project code than simulation code. Another note to discuss in the future.

File: greenheart/tools/eco/hydrogen_mgmt.py [hunk 4]
```diff
@@ -250,15 +342,17 @@ def run_h2_storage(
 
         h2_storage = None
 
-    elif design_scenario["h2_storage_location"] == "turbine":
-        if greenheart_config["h2_storage"]["type"] == "turbine":
+    elif greenheart_config["h2_storage"]["type"] == "turbine":
```
Comment (bayc) [inline greenheart/tools/eco/hydrogen_mgmt.py:345]: Want to consider a different paradigm for this elif structure, thinking along the lines of separate solver files for each type.

File: greenheart/tools/eco/utilities.py [hunk 5]
```diff
@@ -1075,76 +1570,130 @@ def post_process_simulation(
...
+    # save hydrogen information
+    key = "Hydrogen Hourly Production [kg/hr]"
+    np.savetxt(
+        output_dir + "h2_usage",
+        electrolyzer_physics_results["H2_Results"][key],
+        header="# " + key
+    )
 
-    return annual_energy_breakdown
+    return annual_energy_breakdown, hourly_energy_breakdown
```
Comment (bayc) [inline greenheart/tools/eco/utilities.py:1699]: This file seems like a lot of tools, with a handful of utilities.

File: hopp/simulation/hybrid_simulation.py [hunk 6]
```diff
@@ -280,9 +280,21 @@ def __attrs_post_init__(self):
                                                             self.technologies,
                                                             dispatch_options=self.dispatch_options or {})
 
+        # extract o&m costs from the cost_info dict
+        om_cost_info = {}
+        keys_to_remove = []
+        for key in self.cost_info:
+            if "_om_per_kw" in key:
+                om_cost_info.update({key: self.cost_info[key]})
+                keys_to_remove.append(key)
```
Comment (bayc) [inline hopp/simulation/hybrid_simulation.py:289]: Noting for review in the future to reconsider this method.

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 7]
```diff
@@ -160,20 +155,44 @@ def __init__(self,
...
+        else:
+            self.BatterySystem: BatterySystem = BatterySystem()
+
+        if 'system_costs' in fin_config:
+            self.SystemCosts: SystemCosts = SystemCosts.from_dict(fin_config['system_costs'])
+        else:
+            self.SystemCosts: SystemCosts = SystemCosts()
+
+        if 'revenue' in fin_config:
+            self.Revenue: Revenue = Revenue.from_dict(fin_config['revenue'])
```
Comment (bayc) [inline hopp/simulation/technologies/financial/custom_financial_model.py:176]: Noting to convert this class to use attrs in the future so that we can clean up this default setting code.

File: requirements.txt [hunk 8]
```diff
@@ -43,8 +43,8 @@ openpyxl
 CoolProp
 attrs
 utm
-orbit-nrel @ git+https://github.com/WISDEM/ORBIT.git@SemiTaut_Mooring_Update
-pyyaml-include
+orbit-nrel @ git+https://github.com/WISDEM/ORBIT.git@3baed36052d7503de6ea6f2db0b40945d637ea35
+pyyaml-include <= 1.4.1
```
Comment (bayc) [inline requirements.txt:47]: May not need this package with the `load_yaml` utility.

File: examples/analysis/main_usa_new.py [hunk 9]
```diff
@@ -540,8 +541,7 @@ def run_all_hybrid_calcs(site_details, scenario_descriptions, results_dir, load_
         os.mkdir(results_dir)
 
     # directory to resource_files
-    main_dir = Path(__file__).parent.parent.parent
-    resource_dir = main_dir / 'resource_files'
+    resource_dir = f"{ROOT_DIR}/simulation/resource_files/"
```
[the remark this answers is not in the document]
Comment (bayc) [reply examples/analysis/main_usa_new.py]: Updated.

File: examples/greenheart/05-offshore-h2/greenheart-h2-offshore.py [hunk 10]
```diff
@@ -7,15 +7,19 @@
...
 
 PATH = Path(__file__).parent
-YamlIncludeConstructor.add_to_loader_class(loader_class=yaml.FullLoader, base_dir=PATH / './input/floris/')
-YamlIncludeConstructor.add_to_loader_class(loader_class=yaml.FullLoader, base_dir=PATH / './input/turbines/')
+YamlIncludeConstructor.add_to_loader_class(
+    loader_class=yaml.FullLoader, base_dir=PATH / './input/floris/'
+)
+YamlIncludeConstructor.add_to_loader_class(
+    loader_class=yaml.FullLoader, base_dir=PATH / './input/turbines/'
+)
```
[the remark this answers is not in the document]
Comment (bayc) [reply examples/greenheart/05-offshore-h2/greenheart-h2-offshore.py]: Removed.

File: requirements.txt (hunk 8 above)
Comment (bayc) [reply requirements.txt]: Removed.

File: greenheart/simulation/technologies/hydrogen/electrolysis/run_h2_PEM.py (hunk 2 above)
Comment (jaredthomas68) [reply greenheart/simulation/technologies/hydrogen/electrolysis/run_h2_PEM.py:123]: Yes it can.
